content/blog/2019-10-23-AutoVectoring/index.md

```diff
@@ -0,0 +1,242 @@
+
+In terms of implentaion, we found it much easier to first coalesce the loop into one big block before running the strip mine algorithm because we could treat this sequence of instructions as one array instead of having to worry about jumps and label renaming from inserting new blocks. Aggregating all the blocks of this loop was possible because we previously enforced that there can be exactly one branch instruction located at the end of the loop.
+
+Up to here, we have been operating on the assumption that array sizes are divisible by the vector size---four. To account for arrays not divisible by four, we append a copy of this loop (without any optimizations) to a block that follows the main optimized loop. This serially executes the remaining loop iteration which allows us to maintain correctness.
+
+
+### Rust Dynamic Library with Foreign Function Interface
+In order to utilize SIMD intrinsics, we use the Rust crate for SIMD that targets the x86 platform. We design a dynamic library with functions for each of the SIMD operations we support, namely vector-add, vector-multiply, and vector-subtract. These functions are called in the dispatch loop of the interpreter for the corresponding vectorized instructions.
+
+Consider the Rust function for vectorized addition.
+```gherkin=
+#Requires data_a, data_b, and data_c to point to arrays
+# of 32-bit integers of length 4. 
+#Adds the arrays pointed to by data_a and data_b element-wise
+# and store in array pointed to be data_c.
+
+#[no_mangle] 
+pub fn vadd(data_a: *const i32, data_b: *const i32, data_c: *mut i32) {
+    unsafe {
+        let a = _mm_load_si128(mem_a);
+        let b = _mm_load_si128(mem_b);
+        let c = _mm_add_epi32(a, b);
+        _mm_store_si128(mem_c, c);
+        std::mem::forget(mem_c)
+    }
+}
+```
+Consider the corresponding invocation of this Rust function in the interpreter for vector add.
+```gherkin=
+case "vadd": {
+    let vecA = getVec(instr, env, 0);
+    let vecB = getVec(instr, env, 1);
+    let vecC = new Int32Array(fixedVecSize);
+    vadd(vecA, vecB, vecC);
+    env.set(instr.dest, vecC);    
+    return NEXT;
+  }
+```
+These external calls add potentially significant overhead to the execution. We quantify this overhead to more accurately evaluate execution time in the interpreter. We run a vvadd benchmark, documented below, varying array sizes of (128, 1024, 2048, 4196, 8192). We run 10 iterations of each configuration and average the execution time between them. We then compare execution time with and without calls to the Rust library. We find there is a 16% overhead for making Rust calls. While we predict this will be offset for especially large arrays, we decided to add rust calls for serialized instructions as well to isolate the vectorized instructions as a variable in our experiment from the Rust call overhead variable.
```

Comment:
```suggestion
These external calls add potentially significant overhead to the execution. We quantify this overhead to more accurately evaluate execution time in the interpreter. We run a vvadd benchmark, documented below, varying array sizes of (128, 1024, 2048, 4196, 8192). We run 10 iterations of each configuration and average the execution time between them. We then compare execution time with and without calls to the Rust library. We find there is a 16% overhead for making Rust calls. While we predict this will be offset for especially large arrays, we decided to add Rust calls for serialized instructions as well to isolate the vectorized instructions as a variable in our experiment from the Rust call overhead variable.
```